Add [Ordering.O.let=]

snowleopard opened the pull request:
Continuing the conversation started in #5166, this PR adds `let=` to chain comparisons, so one can write:

```ocaml
let compare { x; y; z } t =
  let open Ordering.O in
  let= () = compare_x x t.x in
  let= () = compare_y y t.y in
  compare_z z t.z
```
or, a bit less compactly but more symmetrically:
```ocaml
let compare { x; y; z } t =
  let open Ordering.O in
  let= () = compare_x x t.x in
  let= () = compare_y y t.y in
  let= () = compare_z z t.z in
  Eq
```
The resulting code is pleasant to read but I'm not entirely sure it has no overhead compared to manual `match`ing.

@jeremiedimino Btw, I couldn't use `let?` because `?` isn't an allowed operator character.

P.S.: I converted only a couple of comparison functions but if we decide that going this way is a good idea, I'm happy to adapt the rest of the Dune's code base.

rgrinberg reviewed (approved):
I still think the appeal is limited because of the closure, but I suppose this is harmless.

rgrinberg commented on otherlibs/dune-rpc/private/registry.ml:
```diff
@@ -20,13 +20,11 @@ module Dune = struct
+      let open Ordering.O in
+      let= () = Int.compare t.pid pid in
+      let= () = String.compare t.root root in
+      Where.compare t.where where
```
We should just use polymorphic comparison for simple cases like this.

jberdine replied:
Note though that using Pervasives.compare on a record will make a C call every time while code like that quoted will only make a C call if the pids are equal. Maybe that matters if this is hot and the pid's are often different.

snowleopard replied:
Polymorphic comparisons are a time bomb. They continue working when the underlying data type changes in a way that makes the polymorphic comparison no longer acceptable. Perhaps, annotating every polymorphic comparison with a type helps catch such future changes but it seems better to just avoid polymorphic comparisons altogether.

jberdine commented:
Apologies for the drive-by comment, but another option you might want to consider is to define e.g.
```ocaml
module O = struct
  let map f ord a b = ord (f a) (f b)

  let (>|=) x f = map f x

  let ( @? ) c1 c2 x y =
    match c1 x y with
    | Eq -> c2 x y
    | _ -> c
end
```
and then define compare functions such as:
```ocaml
let compare =
  let open O in
  (compare_x >|= fun u -> u.x)
  @? (compare_y >|= fun u -> u.y)
  @? (compare_z >|= fun u -> u.z)
```
Note that the functions passed to `O.(>|=)` are closed, as is `@?`, so compare functions written in this style can often be inlined down to simple code.

Note that `O.map` and `O.(>|=)` follow the versions in `Containers.Ord`, and `@?` is inspired by `<?>`.

jeremiedimino commented:
Interesting. One additional advantage we get with the `let compare { x; y; z } ...` style is that we get an error if we don't match on all the fields of the record.

I'd be curious how the two approaches compare in terms of generated code though. I would hope that the compiler is able to inline well both versions.

jeremiedimino commented:
I was curious so I ran a quick benchmark:

```ocaml
type ord =
  | Lt
  | Eq
  | Gt

let ( let= ) cmp f =
  match cmp with
  | Eq -> f ()
  | ne -> ne
;;

let map f ord a b = ord (f a) (f b)
let ( >|= ) x f = map f x

let ( @? ) c1 c2 x y =
  match c1 x y with
  | Eq -> c2 x y
  | c -> c
;;

type t =
  { x : int
  ; y : int
  ; z : int
  }

let compare_int (a : int) b = if a < b then Lt else if a = b then Eq else Gt

let compare1 { x; y; z } t =
  let= () = compare_int x t.x in
  let= () = compare_int y t.y in
  let= () = compare_int z t.z in
  Eq
;;

let compare2 =
  (compare_int >|= fun u -> u.x)
  @? (compare_int >|= fun u -> u.y)
  @? (compare_int >|= fun u -> u.z)
;;

let x = { x = 1; y = 1; z = 1 }

let%bench "compare1" =
  ignore (Sys.opaque_identity (compare1 (Sys.opaque_identity x) (Sys.opaque_identity x)))
;;

let%bench "compare2" =
  ignore (Sys.opaque_identity (compare2 (Sys.opaque_identity x) (Sys.opaque_identity x)))
;;
```

And here are the results:

```
Estimated testing time 20s (2 benchmarks x 10s). Change using '-quota'.
┌────────────────────┬──────────┬─────────┬────────────┐
│ Name               │ Time/Run │ mWd/Run │ Percentage │
├────────────────────┼──────────┼─────────┼────────────┤
│ [plop.ml] compare1 │   9.13ns │  11.00w │     30.65% │
│ [plop.ml] compare2 │  29.77ns │         │    100.00% │
└────────────────────┴──────────┴─────────┴────────────┘
```

So the version with `let=` does allocated (:disappointed:) but runs faster. This was with OCaml 4.12, no flambda. The file was compiled with the following options: `-w @a-4-29-40-41-42-44-45-48-52-56-58-59-66-67-68 -strict-sequence -short-paths -strict-formats -g -error-style short -bin-annot -inline 20 -no-alias-deps`.

jeremiedimino commented:
And with flambda:

```
Estimated testing time 20s (2 benchmarks x 10s). Change using '-quota'.
┌────────────────────┬──────────┬────────────┐
│ Name               │ Time/Run │ Percentage │
├────────────────────┼──────────┼────────────┤
│ [plop.ml] compare1 │   6.95ns │    100.00% │
│ [plop.ml] compare2 │   5.13ns │     73.92% │
└────────────────────┴──────────┴────────────┘
```

jberdine commented:
Oh, yeah, I forgot to mention that I only use flambda.

jeremiedimino commented:
@mshinwell do you know why the compiler still allocates a closure when using this `let=` function?

```ocaml
let ( let= ) cmp f =
  match cmp with
  | Eq -> f ()
  | ne -> ne
```

looking at the cmm, `let=` is correctly inlined, but a closure is still systematically allocated for `f`. Given that it's used only once, it feels like it should be inlined.

jeremiedimino commented:
For the sake of the experiment, I also added a manually optimized version to the benchmark:

```ocaml

let compare3 { x; y; z } t =
  match compare_int x t.x with
  | (Lt | Gt) as ne -> ne
  | Eq ->
    (match compare_int y t.y with
     | (Lt | Gt) as ne -> ne
     | Eq -> compare_int z t.z)
```

Here are the results without flambda:

```
$ ../_build/default/plop/inline_benchmarks_runner
Estimated testing time 30s (3 benchmarks x 10s). Change using '-quota'.
┌──────────────────┬──────────┬─────────┬────────────┐
│ Name             │ Time/Run │ mWd/Run │ Percentage │
├──────────────────┼──────────┼─────────┼────────────┤
│ [plop.ml] let=   │   8.30ns │  11.00w │     28.01% │
│ [plop.ml] @?     │  29.63ns │         │    100.00% │
│ [plop.ml] manual │   4.85ns │         │     16.36% │
└──────────────────┴──────────┴─────────┴────────────┘
```

and with flambda:

```
$ ../_build/default/plop/inline_benchmarks_runner
Estimated testing time 30s (3 benchmarks x 10s). Change using '-quota'.
┌──────────────────┬──────────┬────────────┐
│ Name             │ Time/Run │ Percentage │
├──────────────────┼──────────┼────────────┤
│ [plop.ml] let=   │   6.64ns │    100.00% │
│ [plop.ml] @?     │   4.83ns │     72.75% │
│ [plop.ml] manual │   5.74ns │     86.37% │
└──────────────────┴──────────┴────────────┘
```

jeremiedimino commented:
So flambda is great and allows us to write high-level code without guilt, but it's likely that the majority of dune binaries out there are not built with flambda. So we can't rely on flambda at this point.

For this PR, while the micro-benchmarks indicate that `let=` is worse than manually optimized code, it's not clear what is the impact on the whole program. So the next step for this PR would be to push the experiment, convert more comparison functions to this style and observe the impact on the whole binary performances. We use comparison a lot in Dune, so it's reasonable to expect there will be an impact. But who knows, all these allocations are going to be very short lived so maybe the impact is negligible. If the impact is negligible, we might as write use the nicer looking code.

snowleopard commented:
Thanks for the benchmarks @jeremiedimino, and thanks for sharing an interesting alternative @jberdine!

In terms of the code clarity, I prefer the `let=` version. Perhaps, I should add it but with a warning about the overhead, and we would avoid using it for code where performance matters.

jeremiedimino commented:
SGTM. FTR, I already used this function in `sandbox_mode.ml`...

snowleopard commented:
OK, I added a warning to the comment and changed Dune sources to use `let=`. I don't think any of the changes should affect performance. 

I also switched to a more uniform style and got rid of `Poly.compare` in a few places.

The new code style helped me find a couple of places where the existing comparison function was implemented in a dubious (wrong?) way, so I fixed those.

snowleopard commented on src/dune_engine/file_selector.ml:
```diff
@@ -12,10 +12,11 @@ let predicate t = t.predicate
+  let open Ordering.O in
+  let= () = Path.compare dir t.dir in
+  let= () = Predicate.compare predicate t.predicate in
+  Bool.compare only_generated_files t.only_generated_files
```
Here we used to ignore `only_generated_files`.

snowleopard commented on src/dune_engine/package.ml:
```diff
@@ -63,10 +63,10 @@ module Id = struct
-      | s -> s
+    let compare { name; dir } pkg =
+      let open Ordering.O in
+      let= () = Name.compare name pkg.name in
```
Note that here we used to compare package names in the wrong order. This was never noticed because the only place where comparison matters is converting a set to a list and then converting that list back to a different set.

jeremiedimino reviewed (approved):
Looks good. Might be worth comparing the null build of `https://github.com/ocaml-dune/dune-bench` at base and tip of this PR just to check. You can build this repo with:

```
$ cd patdiff.v0.14.0  && dune build --release --print-metrics
```

snowleopard commented:
Btw, here are some results from `./perf.sh`:
```
             Current branch Main branch
             ============== ===========
           | real 5.631s    real 5.642s
Full build | user 29.464s   user 29.530s
           | sys  14.619s   sys  14.842s
                            
           | real 1.092s    real 1.106s
Zero build | user 1.022s    user 1.034s
           | sys  0.071s    sys  0.074s
```

So, just some noise.

snowleopard commented:
@jeremiedimino And here are results when I switch `perf.sh` to use `dune-bench`:

```
             Current branch Main branch
             ============== ===========
           |  real 35.785s   real 35.686s
Full build |  user 170.158s  user 169.040s
           |  sys  40.756s   sys  39.047s

           |  real 1.988s    real 2.021s
Zero build |  user 1.847s    user 1.869s
           |  sys  0.142s    sys  0.152s
```

Again, just noise.

Alizter commented:
I was curious about how the two inline and I did a comparison in assembly.

I have some code like this:
```ocaml
type t2 =
  { x : int
  ; y : int
  ; z : int
  }

let compare1 { x; y; z } t =
  let open O in
  let= () = Int.compare x t.x |> of_int in
  let= () = Int.compare y t.y |> of_int in
  let= () = Int.compare z t.z |> of_int in
  Eq

let compare2 { x; y; z } t =
  match
    ( Int.compare x t.x |> of_int
    , Int.compare y t.y |> of_int
    , Int.compare z t.z |> of_int )
  with
  | ((Lt | Gt) as result), _, _ -> result
  | Eq, ((Lt | Gt) as result), _ -> result
  | Eq, Eq, ((Lt | Gt) as result) -> result
  | Eq, Eq, Eq -> Eq
```
And I checked the assembly output for both using `ocamlopt -S`:

## compare1

```asm
camlOrdering__compare1_301:
	.cfi_startproc
	subq	$8, %rsp
	.cfi_adjust_cfa_offset 8
.L118:
	movq	16(%rax), %rdi
	movq	8(%rax), %rsi
	subq	$48, %r15
	cmpq	(%r14), %r15
	jb	.L119
.L121:
	leaq	8(%r15), %rdx
	movq	%rdx, (%rsp)
	movq	$5367, -8(%rdx)
	movq	camlOrdering__fun_380@GOTPCREL(%rip), %rcx
	movq	%rcx, (%rdx)
	movabsq	$72057594037927941, %rcx
	movq	%rcx, 8(%rdx)
	movq	%rsi, 16(%rdx)
	movq	%rdi, 24(%rdx)
	movq	%rbx, 32(%rdx)
	movq	(%rbx), %rbx
	movq	(%rax), %rdi
	cmpq	%rbx, %rdi
	setl	%al
	movzbq	%al, %rsi
	cmpq	%rbx, %rdi
	setg	%al
	movzbq	%al, %rax
	subq	%rsi, %rax
	leaq	1(%rax,%rax), %rax
	call	camlOrdering__of_int_271@PLT
.L122:
	cmpq	$3, %rax
	je	.L117
	addq	$8, %rsp
	.cfi_adjust_cfa_offset -8
	ret
	.cfi_adjust_cfa_offset 8
	.align	4
.L117:
	movl	$1, %eax
	movq	(%rsp), %rbx
	movq	(%rbx), %rdi
	addq	$8, %rsp
	.cfi_adjust_cfa_offset -8
	jmp	*%rdi
	.cfi_adjust_cfa_offset 8
.L119:
	call	caml_call_gc@PLT
.L120:
	jmp	.L121
	.cfi_adjust_cfa_offset -8
	.cfi_endproc
	.type camlOrdering__compare1_301,@function
	.size camlOrdering__compare1_301,. - camlOrdering__compare1_301
	.text
	.align	16
	.globl	camlOrdering__fun_380
```

## compare2

```asm
camlOrdering__compare2_340:
	.cfi_startproc
	subq	$40, %rsp
	.cfi_adjust_cfa_offset 40
.L139:
	movq	%rax, 8(%rsp)
	movq	%rbx, (%rsp)
	movq	(%rbx), %rbx
	movq	(%rax), %rdi
	cmpq	%rbx, %rdi
	setl	%al
	movzbq	%al, %rsi
	cmpq	%rbx, %rdi
	setg	%al
	movzbq	%al, %rax
	subq	%rsi, %rax
	leaq	1(%rax,%rax), %rax
	call	camlOrdering__of_int_271@PLT
.L140:
	movq	%rax, 16(%rsp)
	movq	(%rsp), %rax
	movq	8(%rax), %rbx
	movq	8(%rsp), %rax
	movq	8(%rax), %rdi
	cmpq	%rbx, %rdi
	setl	%al
	movzbq	%al, %rsi
	cmpq	%rbx, %rdi
	setg	%al
	movzbq	%al, %rax
	subq	%rsi, %rax
	leaq	1(%rax,%rax), %rax
	call	camlOrdering__of_int_271@PLT
.L141:
	movq	%rax, 24(%rsp)
	movq	(%rsp), %rax
	movq	16(%rax), %rbx
	movq	8(%rsp), %rax
	movq	16(%rax), %rdi
	cmpq	%rbx, %rdi
	setl	%al
	movzbq	%al, %rsi
	cmpq	%rbx, %rdi
	setg	%al
	movzbq	%al, %rax
	subq	%rsi, %rax
	leaq	1(%rax,%rax), %rax
	call	camlOrdering__of_int_271@PLT
.L142:
	movq	%rax, %rbx
	movq	16(%rsp), %rax
	cmpq	$3, %rax
	je	.L138
	addq	$40, %rsp
	.cfi_adjust_cfa_offset -40
	ret
	.cfi_adjust_cfa_offset 40
	.align	4
.L138:
	movq	24(%rsp), %rax
	cmpq	$3, %rax
	je	.L137
	addq	$40, %rsp
	.cfi_adjust_cfa_offset -40
	ret
	.cfi_adjust_cfa_offset 40
	.align	4
.L137:
	cmpq	$3, %rbx
	je	.L136
	movq	%rbx, %rax
	addq	$40, %rsp
	.cfi_adjust_cfa_offset -40
	ret
	.cfi_adjust_cfa_offset 40
	.align	4
.L136:
	movl	$3, %eax
	addq	$40, %rsp
	.cfi_adjust_cfa_offset -40
	ret
	.cfi_adjust_cfa_offset 40
	.cfi_adjust_cfa_offset -40
	.cfi_endproc
	.type camlOrdering__compare2_340,@function
	.size camlOrdering__compare2_340,. - camlOrdering__compare2_340
	.data
	.align	8
	.quad	3840
```

The main differences that I see is that compare1 is shorter, but has an extra call to the garbage collector compared to compare2. I can't tell from looking which is faster however.

snowleopard commented:
@Alizter Your `compare2` seems too eager: it will always execute all three `compare`s, right? I think we'd like to avoid that.

Alizter commented:
OK here is a less eager compare3:
```ocaml
let compare3 { x; y; z } t =
  match Int.compare x t.x |> of_int with
  | Eq -> (
    match Int.compare y t.y |> of_int with
    | Eq -> Int.compare z t.z |> of_int
    | result -> result)
  | result -> result
```

## compare3
```asm
camlOrdering__compare3_350:
	.cfi_startproc
	subq	$24, %rsp
	.cfi_adjust_cfa_offset 24
.L145:
	movq	%rax, 8(%rsp)
	movq	%rbx, (%rsp)
	movq	(%rbx), %rbx
	movq	(%rax), %rdi
	cmpq	%rbx, %rdi
	setl	%al
	movzbq	%al, %rsi
	cmpq	%rbx, %rdi
	setg	%al
	movzbq	%al, %rax
	subq	%rsi, %rax
	leaq	1(%rax,%rax), %rax
	call	camlOrdering__of_int_271@PLT
.L146:
	cmpq	$3, %rax
	je	.L144
	addq	$24, %rsp
	.cfi_adjust_cfa_offset -24
	ret
	.cfi_adjust_cfa_offset 24
	.align	4
.L144:
	movq	(%rsp), %rax
	movq	8(%rax), %rbx
	movq	8(%rsp), %rax
	movq	8(%rax), %rdi
	cmpq	%rbx, %rdi
	setl	%al
	movzbq	%al, %rsi
	cmpq	%rbx, %rdi
	setg	%al
	movzbq	%al, %rax
	subq	%rsi, %rax
	leaq	1(%rax,%rax), %rax
	call	camlOrdering__of_int_271@PLT
.L147:
	cmpq	$3, %rax
	je	.L143
	addq	$24, %rsp
	.cfi_adjust_cfa_offset -24
	ret
	.cfi_adjust_cfa_offset 24
	.align	4
.L143:
	movq	(%rsp), %rax
	movq	16(%rax), %rbx
	movq	8(%rsp), %rax
	movq	16(%rax), %rdi
	cmpq	%rbx, %rdi
	setl	%al
	movzbq	%al, %rsi
	cmpq	%rbx, %rdi
	setg	%al
	movzbq	%al, %rax
	subq	%rsi, %rax
	leaq	1(%rax,%rax), %rax
	addq	$24, %rsp
	.cfi_adjust_cfa_offset -24
	jmp	camlOrdering__of_int_271@PLT
	.cfi_adjust_cfa_offset 24
	.cfi_adjust_cfa_offset -24
	.cfi_endproc
	.type camlOrdering__compare3_350,@function
	.size camlOrdering__compare3_350,. - camlOrdering__compare3_350
	.data
	.align	8
	.quad	2044
```

snowleopard commented:
Any interesting conclusions? (Sorry, no time to study the generated code now, putting down some fires.)

Alizter commented:
@snowleopard same as before, the nice code calls GC whereas the handwritten does not. I wonder if all the slowdown comes from the GC call. IIRC there is some OCaml hacker we might be able to do to avoid calling GC there.

snowleopard commented:
To avoid calling GC, we should find a way to stop allocation stuff on the heap (in this case function closures, I think). I suppose, our internal extension where it's possible allocate on the stack instead of the heap might help here, but obviously we won't be able to use it in the external Dune for a long time. I might give it a try internally though!
